fish-rust/src/wildcard.rs

```diff
@@ -11,3 +48,1249 @@ pub const ANY_STRING_RECURSIVE: char = char_offset(WILDCARD_RESERVED_BASE, 2);
 /// This is a special pseudo-char that is not used other than to mark the
 /// end of the the special characters so we can sanity check the enum range.
 pub const ANY_SENTINEL: char = char_offset(WILDCARD_RESERVED_BASE, 3);
+
+#[derive(PartialEq)]
+pub enum WildcardResult {
+    /// The wildcard did not match.
+    NoMatch,
+    /// The wildcard did match.
+    Match,
+    /// Expansion was cancelled (e.g. control-C).
+    Cancel,
+    /// Expansion produced too many results.
+    Overflow,
+}
+
+fn resolve_description<'f>(
+    full_completion: &wstr,
+    completion: &mut &wstr,
+    expand_flags: ExpandFlags,
+    description_func: Option<&'f dyn Fn(&wstr) -> WString>,
+) -> WString {
+    if let Some(complete_sep_loc) = completion.find_char(PROG_COMPLETE_SEP) {
+        // This completion has an embedded description, do not use the generic description.
+        assert!(
+            completion.len() > complete_sep_loc,
+            "resolve_description lacks a length assumption"
+        );
+        let (comp, description) = completion.split_at(complete_sep_loc + 1);
+        *completion = comp;
```

Comment:
Testing revealed there's an off-by-one error in this code; "completion" should not include
the separator. I changed it to

        let description = completion[complete_sep_loc + 1..].to_owned();
        *completion = &completion[..complete_sep_loc];